x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/utils/persistence/ResultsPersisterService.java

```diff
@@ -208,48 +193,195 @@ private static boolean isIrrecoverable(Exception ex) {
         return IRRECOVERABLE_REST_STATUSES.contains(status(t));
     }
 
-    /**
-     * {@link RetryContext} object handles logic that is executed between consecutive retries of an action.
-     *
-     * Note that it does not execute the action itself.
-     */
-    private class RetryContext {
+    @SuppressWarnings("NonAtomicOperationOnVolatileField")
+    private static class BulkRequestRewriter {
```

Comment:
I'm not convinced this class is bringing much to the party. 

`private volatile BulkRequest bulkRequest;` could be a field in `BulkRetryableAction` and `buildNewRequestFromFailures` could be called directly.

Reply:
>could be a field in BulkRetryableAction and buildNewRequestFromFailures could be called directly.

No, it cannot. Because the action is passed via `super` the `super` is ctor'd before the child class is. So, references DIRECTLY to the child class methods is not possible.

This class guaruntees things are constructed and mutable inside the lambda passed to super.